cscs-checks/prgenv/affinity_check.py

```diff
@@ -3,184 +3,547 @@
-                'ntasks_per_core': 1,
-                'multithreading': None,
-                'OMP_PLACES': 'cores',
-            },
-        }
-        self.variant = variant
+        '''Hook to control multithreading settings for each system.'''
+
+        cp = self.current_partition.fullname
+        mthread = (
+            self.system.get(cp, {}).get('multithreading', None) or
+            self.multithread
+        )
+        if mthread:
+            self.use_multithreading = mthread
 
     @rfm.run_before('run')
-    def set_tasks_per_core(self):
-        partname = self.current_partition.name
-        self.num_cpus_per_task = (
-            self.cases[self.variant]['num_cpus_per_task:%s' % partname])
-        if self.cases[self.variant]['ntasks_per_core']:
-            self.num_tasks_per_core = (
-                self.cases[self.variant]['ntasks_per_core'])
+    def set_launcher(self):
+        '''Hook to control hints and cpu-bind for each system.'''
 
+        cp = self.current_partition.fullname
+        cpu_bind = (
+            self.system.get(cp, {}).get('cpu-bind', None) or self.cpu_bind
+        )
+        if cpu_bind:
+            self.job.launcher.options += [f'--cpu-bind={cpu_bind}']
+
+        hint = self.system.get(cp, {}).get('hint', None) or self.hint
+        if hint:
+            self.job.launcher.options += [f'--hint={hint}']
+
+
+class AffinityOpenMPBase(AffinityTestBase):
+    '''Extend affinity base with OMP hooks.
+
+    The tests derived from this class book the full node and place the
+    threads acordingly based exclusively on the OMP_BIND env var. The
+    number of total OMP_THREADS will vary depending on what are we
+    binding the OMP threads to (e.g. if we bind to sockets, we'll have as
+    many threads as sockets).
+    '''
+
+    omp_bind = variable(str)
+    omp_proc_bind = variable(str, value='spread')
+
+    def __init__(self):
+        super().__init__()
         self.num_tasks = 1
+
+    @property
+    def ncpus_per_task(self):
+        '''We use this property to set the hook below and keep exec order.'''
+        return self.num_cpus
+
+    @rfm.run_before('run')
+    def set_num_cpus_per_task(self):
+        self.num_cpus_per_task = self.ncpus_per_task
+
+    @rfm.run_before('run')
+    def set_omp_vars(self):
         self.variables = {
-            'OMP_NUM_THREADS': str(self.num_cpus_per_task),
-            'OMP_PLACES': self.cases[self.variant]['OMP_PLACES']
-            # OMP_PROC_BIND is set to TRUE if OMP_PLACES is defined.
-            # Both OMP_PROC_BIND values CLOSE and SPREAD give the same
-            # result as OMP_PROC_BIND=TRUE when all cores are requested.
+            'OMP_NUM_THREADS': str(self.num_omp_threads),
+            'OMP_PLACES': self.omp_bind,
+            'OMP_PROC_BIND': self.omp_proc_bind,
         }
 
+    @rfm.run_before('sanity')
+    def consume_cpu_set(self):
+        raise ValueError('this function must be overridden')
```

Comment:
If this is meant to act as an abstract method, better raise a `NotImplementedError`.